System V shared memory APIs (2nd draft)

IzawGithub opened the pull request:
## What does this PR do

## Checklist:

- [X] I have read `CONTRIBUTING.md`
- [ ] I have written necessary tests and rustdoc comments
- [X] A change log has been added if this PR modifies nix's API

I've seen issue #1718 and the pull request #1989 made by @arpankapoor from a year ago, that added some support for the System V API.

This is a continuation of this PR, where I tried to implement most of the change suggested by @VorpalBlade.

## What's missing

I haven't added any unit test yet, as somehow I'm getting a lot of compilation error on the test, about some missing symbol, but I've got a closed source project where I'm using this draft as a "eat your own dog food", with unit test, and it seems to work fine.

Some of the bitflags, 2 in `ShmgetFlag` and a lot in `SemctlCmd` do not exist in libc, even though they are documented.
So to merge this, there is a need to create a PR for libc that add those.

## What can be better

`Permissions`: Could maybe be moved elsewhere? I haven't found any struct already implemented that wrap the octal permission system unix use, so I threw this in 5m.

`shmat`: I'm **really** not a fan of returning a void ptr. I've tried to wrap it in a `Weak<RefCell<T>>` but I get a segfault when it gets dropped.
What would be ideal is for the user to give a generic type T that could be wrapped in a non owning smart pointer that could also be null. That's why I though Weak would be the answer, but I'm not good enough to find the problem.
Currently, the user has the cast it in unsafe themselves to be able to use the value in it, which is awkward and error prone.

I'm not too sure about the linux only bitflags compilation. We should test it for android too, but I don't know how to do that.

SteveLauC commented:
I am trying to add support for System V message queue in #2318

IzawGithub commented:
Hello @SteveLauC,

I think this PR is pretty good now. I've added some basic test, with around 30% coverage of the new function.
What's missing is testing `shmdt`, which I don't really know how I could do that, and all of the different bitflags, which is a bit overwhelming.

I also didn't implement the `Semop` function because I have no clue at what it's even supposed to do. If anyone got an idea about how to implement it, it would be more than welcome.

I don't understand why BSD isn't able to build successfully, when it should be working fine.
<br>

Is there anything left to be done before a review and an eventual merge?

SteveLauC commented:
Hi @IzawGithub, sorry for the late reply, I plan to take a look at this PR this weekend :)

SteveLauC commented on src/sys/system_v.rs:
```diff
@@ -0,0 +1,385 @@
+pub fn shmget(
+    key: key_t,
+    size: size_t,
+    shmflg: Vec<ShmgetFlag>,
```
Do we need to use a `Vec` here, considering that `ShmgetFlag` is a bitflag so that it is ORable, for example, you can do:

```rs
shmget(key, size, ShmgetFlag::IPC_CREAT | ShmgetFlag::IPC_EXCL, permission)

IzawGithub replied:
You're right, I forgot about that.
Fixed by changing to taking a `Flag` directly

SteveLauC commented on src/sys/mod.rs:
```diff
@@ -143,6 +143,10 @@ feature! {
 #[allow(missing_docs)]
 pub mod sysinfo;
 
+#[allow(missing_docs)]
```
I would appreciate it if these new APIs have nice documents or even better, examples:)

IzawGithub replied:
Fixed, removed the allow lint and added some doctest, please tell me if anything isn't clear in it.

SteveLauC commented on src/sys/mod.rs:
```diff
@@ -143,6 +143,10 @@ feature! {
 pub mod sysinfo;
 
+#[allow(missing_docs)]
+#[cfg(any(bsd, target_os = "linux", not(target_os = "android")))]
```
And it is always ok to not support all the platforms, we can add them incrementally

IzawGithub replied:
Fixed the redundant `android` target. BSD is still not building and I don't really understand why. Something to do with macro? I'm not proficient with them yet.

SteveLauC replied:
> BSD is still not building and I don't really understand why.

The OpenBSD and NetBSD CI failed because some missing constants in the [libc](https://github.com/rust-lang/libc) crate, which means we need to add them before merging this PR if we want to have `#[cfg(bsd)`s supported.

Let me take a closer look at this and probably give it a fix!

SteveLauC commented on changelog/2313.added.md:
```diff
@@ -0,0 +1 @@
+Add SystemV IPC support
```
It seems that this PR only covers:

1. semaphore set
2. shared memory segments

so we should probably annotate this

IzawGithub replied:
Kinda fixed, I'm not sure if I should put more detail in it?

SteveLauC replied:
> I'm not sure if I should put more detail in it?

No need for details, a brief intro will be sufficient.

SteveLauC commented on src/sys/system_v.rs:
````diff
@@ -0,0 +1,385 @@
+/// # Ok(())
+/// # }
+/// ```
+pub struct Permissions {
````
For the last 9 bits of those `flg` arguments, Nix already has a type for it,
   `nix::sys::stat::Mode`, so a new type is probably not necessary.

IzawGithub replied:
Fixed, wasn't aware of this type. This makes it depend on the `fs` module but I'm assuming it's fine?

SteveLauC replied:
>  This makes it depend on the fs module but I'm assuming it's fine?

Yeah, this is unfortunate, ideally, this `Mode` type should be available if `#[cfg(any(feature="fs", feature="system_v")]`, doing this will require some extra work, so I think it is ok to make `system_v` depend on `fs`.

SteveLauC commented:
Some thoughts about these APIs:

1. Module names

   I kinda think we should separate `semaphore set` and `shared memory segments`
   into 2 modules, following the names of their C header files, i.e, `sys/sem.rs` for 
   `semaphore set` and `sys/shm.rs` for `shared memory segments`.

2. For those APIs, do you think it will be better if we do something like:

   Take `semaphore set` as an example:

   ```rs
   /// A System V semaphore set.
   pub struct SemaphoreSet(libc::c_int);

   pub enum SemCmd {
       IPC_STAT(&mut libc::semid_ds),
       IPC_SET(&libc::semid_ds),
   }

   impl SemaphoreSet {
       pub fn new(key: key_t, n_sem: usize, semflg: SemFlg) -> Result<Self>;
       pub fn op<O: AsRef<[Sembuf]>>(&self, ops: O) -> Result<()>;
       pub fn op_timeout<O: AsRef<[Sembuf]>>(&self, ops: O, timeout: TimeSepc) -> Result<()>;
       pub fn ctl(&self, sem_num: usize, cmd: SemCmd) -> Result<Option<libc::c_int>>;
   }
   ```

   And to remove a `semaphore set`, I am thinking about making our wrapper type 
   non-Copy, and do this:

   ```rs
   #[allow(missing_copy_implementations)]
   pub struct SemaphoreSet(libc::c_int);

   impl SemaphoreSet {
       pub fn remove(self) -> Result<(), (Errno, Self)> {
          let res = unsafe {
              libc::semctl(self.0, /*whatever*/, IPC_RMID)
          };

          match Errno::result(res) {
              Ok(_) => Ok(()),
              Err(errno) => Err((errno, self))
          }
       }
   }
   ```

IzawGithub commented:
Hello,

Thanks for the comprehensive review! Those are all fair issue, I'll work on them this week.

For your last question about wrapping all the functions, this is certainly doable, but I wasn't sure if it was prefered to match libc as closely as possible or not. I can definitely try to make something more Rusty. Worst case scenario, if it's not ok we could always revert to what's currently there.

Also, once the PR is done, should I clean the history with a rebase before a merge or not?

SteveLauC commented:
> Also, once the PR is done, should I clean the history with a rebase before a merge or not?

Nix uses squash merge, so it's fine to leave the commit history here:)

IzawGithub commented:
Hello @SteveLauC, sorry for not answering sooner, I was pretty sick.

I have refactored the shared memory segment portion of this PR, to fit more with Rust style of doing things. If you feel that this is the right direction to go, I'll work on making something similar for the semaphore portion.

This also resolve the issue of returning a raw pointer. Before, the end user had to use `unsafe` for deref. This is now fixed because `SharedMemory` acts like a smart pointer to the non owning data. This makes for a much nicer interface.

IzawGithub commented on changelog/2313.added.md:
Fixed

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+    ///
+    pub fn new(
+        shmid: c_int,
+        shmaddr: Option<c_void>,
```
It seems that one cannot create a variable of type `c_void`? (unless with `std::mem::transmute()`) Which means that the case `Some(a_value_of_type_c_void)` will never happen?

IzawGithub replied:
Huh, I didn't think about that. I don't really see a way to fix it. Most use I'm finding online of `shmat` is with shmaddr set to null.
So maybe not having that feature is ok? If I remove it, I'll add it to the doc that this isn't supported.

SteveLauC replied:
I hope we can have it if possible as there will always be someone who wants to use it, so if we are going to keep this argument, what is the proper type? 

1. `Option<&libc::c_void>`
  
     If `SHM_RND` is not specified in the flag argument, then the address must be page-aligned, which means in most cases, users need to manually specify a page-aligned address, and reference, is rarely page-aligned, so it is not that useful.

2. `Option<*const libc::c_void>`
    
    `Option` is rusty, but if the inner type is a raw pointer, why do we need it?

3. a simple raw pointer `*const libc::c_void`
   
   Simple, unrusty, but work.

Your thoughts on this?

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+}
+
+impl<T> SharedMemory<T> {
+    /// Create a new SharedMemory object
```
This should probably removed considering that it won't create a new segment but attach to an existing one.

IzawGithub replied:
Fixed by removing the ctor and letting `Shm` wraps around it.

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+        shmid: c_int,
+        shmaddr: Option<c_void>,
+        shmat_flag: ShmatFlag,
+        mode: Mode,
```
Looks like that the `shmflg` argument of `shmat(2)` should not contain the permission info

IzawGithub replied:
Oops, yeah, copy and paste error, my bad. Fixed

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+    pub fn shmctl(
+        &self,
+        shmctl_flag: ShmctlFlag,
+        buf: Option<shmid_ds>,
```
What about passing it by reference:

```suggestion
        buf: Option<&shmid_ds>,
```

IzawGithub replied:
Yes, it's even transformed to a pointer after. Fixed

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+    /// [`shmctl(2)`]: https://man7.org/linux/man-pages/man2/shmctl.2.html
+    pub fn shmctl(
+        &self,
+        shmctl_flag: ShmctlFlag,
```
This name can be confusing, I think we should stick with the C variable name, i.e., something like `cmd` or `shm_cmd`.

IzawGithub replied:
No problem, renamed to `shm_cmd`

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+        &self,
+        shmctl_flag: ShmctlFlag,
+        buf: Option<shmid_ds>,
+        mode: Mode,
```
Same here, the `shmctl(2)` syscall does not seem to take a `mode` argument

IzawGithub replied:
Fixed.

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+    /// [`shmat(2)`]: https://man7.org/linux/man-pages/man2/shmat.2.html
+    fn shmat(
+        shmid: c_int,
+        shmaddr: Option<c_void>,
```
Same here, see above

IzawGithub replied:
Fixed.

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,448 @@
+        shmid: c_int,
+        shmaddr: Option<c_void>,
+        shmat_flag: ShmatFlag,
+        mode: Mode,
```
Same here, `mode` argument seems to be unneeded

IzawGithub replied:
Fixed.

IzawGithub commented:
Hello, all of the issue listed above should be fixed, expect  for the `c_void` where I don't really know how to go forward with.

Nice catch on the UB!
To resolve it, I had to make `shmget` an unsafe function. This is because, if the user is connecting to a shared memory zone that we haven't created ourselves, then there is no way to make sure the types match.
If we're always creating the new memory zone, then we're sure the type match, so I've added a safe `create_and_connect` for that.

I'll work on the semaphores during the coming week.

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+
+impl<T> SharedMemory<T> {
+    /// Performs control operation specified by `cmd` on the System V shared
+    /// memory segment given by `shmid`.
```
This `shmid` identifier is not a parameter but stored in `SharedMemory`

IzawGithub replied:
Fixed

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+#[derive(Debug)]
+/// Safe wrapper around a SystemV shared memory segment data
+///
+/// This is a smart pointer, and so implement the [`Deref`] and [`DerefMut`] traits.
```
Nit: seems to be a mistake:

```suggestion
/// This is a smart pointer, and so implements the [`Deref`] and [`DerefMut`] traits.
```

IzawGithub replied:
Fixed

SteveLauC commented on src/sys/system_v/shm.rs:
````diff
@@ -0,0 +1,476 @@
+    /// ```
+    ///
+    /// [`shmctl(2)`]: https://man7.org/linux/man-pages/man2/shmctl.2.html
+    pub fn shmctl(
````
Should we move this method to `impl Shm { }`, it is more like a method associated with `Shm`

SteveLauC replied:
Yeah, I think this should be done or users will have to attach to a shared memory segment in order to remove it, which is not needed in the C API

IzawGithub replied:
This is even better, that way SharedMemory really is just a smart pointer and has no associated function except the private detach. Fixed.

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+    /// Valid flags for the third parameter of the function [`shmget`]
+    pub struct ShmgetFlag: c_int
+    {
+        /// A new shared memory segment is created if key has this value.
+        IPC_PRIVATE;
```
If this is a key rather than a value of flag, then we should probably remove it from `ShmgetFlag`?

IzawGithub replied:
After re-reading the doc, you're right. This isn't a bitflag but a value the user can pass to the `key` argument. Fixed.

SteveLauC commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+}
+
+libc_bitflags!(
+    /// Valid flags for the third parameter of the function [`shmget`]
```
This makes it render correctly:

```suggestion
    /// Valid flags for the third parameter of the function [`Shm::shmget`].
```

IzawGithub replied:
Forgot to edit that, nice catch. Fixed.

IzawGithub commented:
Hello, this is harder than I though!

For your example, after a bit of debugging, I think that pointing to 0x0 is because we're trying to store a type that's a ref, when SharedMemory actually expect a non ref.
`T` being a generic, I would like to restrict it to only owned data, which would resolve the issue.
However, I don't really see a way to do that. I didn't see any "not a pointer" trait, maybe `Default`? But that's annoying for the end user if their data struct does not implement default.

The rest of the issue are pretty simple fix I'll work on right away.

IzawGithub commented on src/sys/system_v/shm.rs:
````diff
@@ -0,0 +1,476 @@
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    pub fn create_and_connect(key: key_t, mode: Mode) -> Result<Self> {
````
Yeah I didn't really put too much though on the name. Maybe `create_and_get`? We're creating the segment by using shm*get*. I'm not the best at naming things, so if you've got a better idea, feel free!

IzawGithub commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+    /// If you need more customisation, use the unsafe version,
+    /// [`Shm::shmget`], with the key [`ShmgetFlag::IPC_CREAT`].
+    ///
+    /// Attaching a segment to a specific adress isn't supported. This is
+    /// because there is no way to create a void pointer on rust.
```
I was worried of a potential deref of a null pointer, but I remembered that the libc shmat already handle that, so this is actually fine and not that bad to use as a end user. Added

IzawGithub commented on src/sys/mod.rs:
```diff
@@ -143,6 +143,12 @@ feature! {
+#[cfg(any(bsd, target_os = "linux"))]
+feature! {
+    #![feature = "system_v"]
+    pub mod system_v;
```
Oh I didn't realise that this would all go in `sys`, I though you meant a sysv specific module, my bad. Fixed

IzawGithub commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+        /// segment is attached for read and write access, and the process must
+        /// have read and write permission for the segment.
+        /// There is no notion of a write-only shared memory segment.
+        SHM_RDONLY;
```
Yeah, it was there before I removed it with the shmat pointer adress var. Added back

IzawGithub commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+        /// segment. When swap space is not reserved one might get SIGSEGV upon
+        /// a write if no physical memory is available. See also the discussion
+        /// of the file /proc/sys/vm/overcommit_memory in proc(5).
+        #[cfg(linux)]
```
No problem, fixed.

IzawGithub commented on Cargo.toml:
```diff
@@ -64,6 +64,7 @@ resource = []
 sched = ["process"]
 signal = ["process"]
 socket = ["memoffset"]
+system_v = ["fs"]
```
Sure, fixed

IzawGithub commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+);
+
+#[cfg(test)]
+mod tests {
```
I used private field before, but not anymore, so this can indeed be an independent test. Fixed.

IzawGithub commented on src/sys/system_v/shm.rs:
```diff
@@ -0,0 +1,476 @@
+        let size = std::mem::size_of::<T>();
+        // This is the main difference between this function and [`Shm::shmget`]
+        // Because we are always creating a new segment, we can be sure that the size match
+        let shmget_flag = ShmgetFlag::IPC_CREAT | ShmgetFlag::IPC_EXCL;
```
I do not understand what you mean here.
Why would `.attach()` use `shmget_flag`, when it only connect to an existing memory segment?

This here is guaranteed safe because we're 100% sure that the new memory segment either has the size of `T`, or if it fails to be created, return an error. If something more fine grained is needed, isn't `unsafe .shmget()` the function that fulfill this purpose?

SteveLauC commented:
Hi, sorry for the late response, been busy with my work and can barely do open source things on work days.

-----

> For your example, after a bit of debugging, I think that pointing to 0x0 is because we're trying to store a type that's a ref, when SharedMemory actually expect a non ref. T being a generic, I would like to restrict it to only owned data, which would resolve the issue.

Even if we restrict `T` to be owned types, it is still not guaranteed that the value stored in that shared memory is valid for T, so the problem still exists.

From my observation, that memory segment will be zero-initialized, 0 is a valid type for `std::num::NonZeroU8`:

```rs
use nix::sys::{
    stat::Mode,
    system_v::shm::{Shm, ShmatFlag},
};

fn main() {
    let mode = Mode::S_IRWXU;

    let shm = Shm::<std::num::NonZeroU8>::create_and_connect(65536, mode).unwrap();
    let mem_segment = shm.attach(ShmatFlag::empty()).unwrap();
    println!("number = {}", *mem_segment);
}
```

```sh
$ cargo r                      
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s
     Running `target/debug/rust`
number = 0
```
